openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py

```diff
@@ -0,0 +1,520 @@
+    def test_process_request_timeout_exceeded_logs_out_user(self, mock_auth, mock_datetime):
+        """
+        Test that request exceeding timeout period logs out user.
+        """
+        # Set up existing timestamp (10 minutes ago)
+        last_touch = datetime(2025, 6, 16, 12, 0, 0)
+        current_time = datetime(2025, 6, 16, 12, 10, 0)  # 10 minutes later
+        
+        self.request.session[LAST_TOUCH_KEYNAME] = last_touch.isoformat()
+        
+        # Mock all datetime.utcnow() calls to return current_time
+        mock_datetime.utcnow.return_value = current_time
+        mock_datetime.fromisoformat = datetime.fromisoformat
+        
+        response = self.middleware.process_request(self.request)
+        
+        assert response is None
+        # The middleware deletes the key before logging out
+        assert LAST_TOUCH_KEYNAME not in self.request.session
+        mock_auth.logout.assert_called_once_with(self.request)
+
+    @override_settings(SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=300)
+    @patch('openedx.core.djangoapps.session_inactivity_timeout.middleware.datetime')
+    def test_process_request_invalid_timestamp_handles_gracefully(self, mock_datetime):
+        """
+        Test that invalid timestamp in session is handled gracefully.
+        """
+        # Set up invalid timestamp
+        self.request.session[LAST_TOUCH_KEYNAME] = "invalid-timestamp"
+        mock_now = datetime(2025, 6, 16, 12, 0, 0)
+        mock_datetime.utcnow.return_value = mock_now
+        
+        response = self.middleware.process_request(self.request)
+        
+        assert response is None
+        # Should set new timestamp despite invalid old one
+        assert self.request.session[LAST_TOUCH_KEYNAME] == mock_now.isoformat()
+
+    @override_settings(
+        SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=900,  # 15 minutes timeout
+        SESSION_ACTIVITY_SAVE_DELAY_SECONDS=600  # 10 minutes save frequency
+    )
+    @patch('openedx.core.djangoapps.session_inactivity_timeout.middleware.datetime')
+    def test_process_request_session_save_frequency_within_period(self, mock_datetime):
+        """
+        Test that session.modified is set to False when within save frequency period.
+        """
+        # Set up times
+        last_save = datetime(2025, 6, 16, 12, 0, 0)
+        current_time = datetime(2025, 6, 16, 12, 5, 0)  # 5 minutes later (< 10 min frequency, < 15 min timeout)
+        
+        self.request.session[LAST_TOUCH_KEYNAME] = last_save.isoformat()
+        self.request.session[LAST_SESSION_SAVE_TIME_KEYNAME] = last_save.isoformat()
+        
+        # Mock datetime methods properly
+        mock_datetime.utcnow.return_value = current_time
+        mock_datetime.fromisoformat = datetime.fromisoformat
+        
+        response = self.middleware.process_request(self.request)
+        
+        assert response is None
+        assert self.request.session.modified is False
```

Comment:
1. I'm not going to review unit tests in-depth until the logic gets fixed to avoid marking the session modified as False.
2. Consider using exploring `@ddt.ddt`, `@ddt.data`, `ddt.unpack` (see other tests as examples) to avoid some test redundancy.